pctasks/core/pctasks/core/utils/summary.py

```diff
@@ -0,0 +1,673 @@
+"""summary.py
+
+Module for generating summaries from JSON objects.
+Useful for generating Collection "summaries" and "item-assets" fields.
+"""
+
+
+from abc import abstractmethod
+from enum import Enum
+from functools import reduce
+from typing import Any, Dict, List, Optional, Set, Union, cast
+
+from pydantic import BaseModel, Field
+
+from pctasks.core.utils import map_opt
+
+
+class SummarySettings(BaseModel):
+    max_mixed_summary_samples = 4
+    max_mixed_value_samples = 4
+    max_mixed_object_list_samples = 3
+    max_mixed_key_sets = 5
+
+
+class ValueTypes(str, Enum):
+    BOOLEAN = "boolean"
+    INT = "numeric-int"
+    FLOAT = "numeric-float"
+    STRING = "string"
+    LIST = "list"
+    NULL = "null"
+
+
+class SummaryTypes(str, Enum):
+    DISTINCT = "distinct"
+    INT_RANGE = "int-range"
+    FLOAT_RANGE = "float-range"
+    OBJECT_LIST = "object-list"
+    MIXED_VALUE = "mixed-value"
+    MIXED_OBJECT_LIST = "mixed-object-list"
+    MIXED_SUMMARIES = "mixed-summaries"
+
+
+def is_value_type(v: Any) -> bool:
+    return (
+        isinstance(v, bool)
+        or isinstance(v, str)
+        or isinstance(v, float)
+        or isinstance(v, int)
+    )
+
+
+class ValueCount(BaseModel):
+    type: str
+    count: int
+
+    def add_count(self, other: "ValueCount") -> None:
+        self.count += other.count
```

Comment:
Is the method name `add_count` meaningful? Or can you use `__add__` and then just call it with `+`?

And I haven't looked at the context yet, but it feels a bit strange to have a class for this. Can we just use an int? Or maybe a `collections.Counter`? (I don't know what `type` is yet).

Reply:
`type` describes the type of the value, so having a Pydantic model that defines a type that base classes overrides with literal values lets Pydantic deserialize the classes correctly. The setup of the classes has a lot to do with how Pydantic serializes/deserializes based on type info.

This could use plus, but I will often prefer non-magic methods for the sake of being explicit, at least for internal APIs - as a reader I might be confused on whether it was a custom class or just an int if I saw a +.

Reply:
Renaming to `merge`, that may be less confusing as it fits with the naming of other classes.